src/OmniSharp.Roslyn.CSharp/Services/Completion/CompletionService.cs

```diff
@@ -498,16 +207,60 @@ public async Task<CompletionResolveResponse> Handle(CompletionResolveRequest req
             };
         }
 
-        private static LinePositionSpanTextChange GetChangeForTextAndSpan(string? insertText, TextSpan changeSpan, SourceText sourceText)
+        public async Task<CompletionAfterInsertResponse> Handle(CompletionAfterInsertRequest request)
         {
-            var changeLinePositionSpan = sourceText.Lines.GetLinePositionSpan(changeSpan);
-            return new()
+            var cachedList = _cache.GetCachedCompletionList(request.Item.Data.CacheId);
+            if (cachedList is null)
+            {
+                _logger.LogError("Cannot call completion/afterInsert before calling completion!");
+                return new CompletionAfterInsertResponse();
+            }
+
+            var (_, document, _, completions) = cachedList;
+            var index = request.Item.Data.Index;
+
+            if (request.Item is null
+                || index >= completions.Items.Length
+                || index < 0
+                || request.Item.TextEdit is null)
+            {
+                _logger.LogError("Received invalid completion afterInsert!");
+                return new CompletionAfterInsertResponse();
+            }
+
+            var lastCompletionItem = completions.Items[index];
+            if (lastCompletionItem.DisplayTextPrefix + lastCompletionItem.DisplayText + lastCompletionItem.DisplayTextSuffix != request.Item.Label)
+            {
+                _logger.LogError("Inconsistent completion data. Requested data on {0}, but found completion item {1}", request.Item.Label, lastCompletionItem.DisplayText);
+                return new CompletionAfterInsertResponse();
+            }
+
+            if (lastCompletionItem.GetProviderName() is not (CompletionItemExtensions.OverrideCompletionProvider or
+                                                             CompletionItemExtensions.PartialMethodCompletionProvider)
+                                                        and var name)
```

Comment:
oh I didn't know you could do something like this 🧐

Reply:
Composition!